### PR TITLE
Improved program download, multi-file projects, and persistent storage on Powered Up hubs

### PR DESCRIPTION
**Pull request by laurensvalk**
This is a major upgrade on how we load, keep, and save user scripts. Highlights:

- Non-blocking system-file access on SPIKE hubs. This is a work in progress. Programs will load faster, and it should fix https://github.com/pybricks/support/issues/679. It also works around compatibility issues with file systems of other firmwares.
- Migration of the program download process from MicroPython to pbio. This will enable us to improve the download protocol later on. For now it remains compatible with Pybricks Code and Pybricksdev.
- Saving programs on Move Hub, Technic Hub, and City Hub without reinstalling the firmware.
- Improved RAM stack/heak configuration, leaving more RAM for user programs and data.

----

Until now, we had to discard the downloaded user program to avoid using twice the amount of RAM when running it. This PR explores an experimental MicroPython feature (https://github.com/micropython/micropython/pull/8381) that lets us run the program without making a full copy.

This means we can keep the program in RAM as we run it. In turn, this means we can run it again by just pressing the button, and also leave more RAM for user variables. But perhaps more importantly, keeping the program means we can save it *on the hub* on shutdown. We could not do this before because flash access is a "slow" process that blocks other things like sensor communication. But if we're shutting down anyway, that doesn't matter. This also considerably reduces the number of times you'd save something on flash.

On SPIKE hubs, discarding the program from RAM was not an issue because it could be saved on the external file system. However, this PR gives us the opportunity to solve a related problem. There are several (official and unofficial) firmwares out there for SPIKE hubs. When each firmware uses its own file system, things get messy and it is hard to explain to new users. We can work around this by storing the user program in a region of the external flash that isn't used by the file system. Programs can only be as large as the RAM (~280Kb), so not much space is needed. If we ever enable user access to the file system, we could do it the MicroPython way as long as the low-level I/O operations are not blocking.

I expect that this will not be merged soon. It will need a thorough review and extensive testing, but this should allow for some discussion and interactive review.

The Move Hub build size increases by 868 bytes, which is not bad considering what is being added. There should be room for quite a bit of optimization as well.

---

Some remaining tasks:

- [x] Make SPI file access work properly on the SPIKE Essential Hub. Consider using DMA for data transfers.
- [ ] Generalize _stm32_ port to _embedded_, so we can use all the same code on boards like the NXT as well.
- [ ] Enable user access to small memory space for saving persistent settings.
- [x] Review the MicroPython patch, which currently has some tests failing.
- [ ] Review the required stack size on all hubs so we leave more to the MicroPython heap if we can.
- [ ] Review the amount of persistent storage on Powered Up hubs. On Move Hub, for example, it is currently 2K but we could allow 4K or more. There is currently 9K free for this but we should reserve some space for future features.
- [ ] Extensive testing on all hubs.
- [x] Revert bluetooth hacks that were used to debug shutdown sequence.
- [x] Revisit MicroPython import hook to match preferred method suggested in https://github.com/micropython/micropython/pull/9026

Revisions:

- [x] Merge spi driver into block device instead of generic approach

### CODE REVIEW

File: lib/pbio/sys/program_load.c [hunk 1]
```diff
@@ -0,0 +1,283 @@
...
+        etimer_set(&timer, 500);
+
+        // Size of chunk to receive now, and location of that chunk.
+        chunk_size = remaining_size < PBSYS_CONFIG_DOWNLOAD_CHUNK_SIZE ?
+            remaining_size : PBSYS_CONFIG_DOWNLOAD_CHUNK_SIZE;
+        chunk_buf = map->program_data + incoming_size - remaining_size;
+
+        // Receive a chunk of data or stop on timeout. The chunk may be further
+        // broken down by the sender, so receive it piece by piece.
+        etimer_set(&timer, 500);
```
Comment (dlech) [inline lib/pbio/sys/program_load.c:181]: timer is set twice before waiting

File: lib/pbio/test/pbsysconfig.h [hunk 2]
```diff
@@ -4,3 +4,5 @@
 #define PBSYS_CONFIG_BLUETOOTH                      (1)
 #define PBSYS_CONFIG_HUB_LIGHT_MATRIX               (1)
 #define PBSYS_CONFIG_STATUS_LIGHT                   (1)
+#define PBSYS_CONFIG_PROGRAM_LOAD                   (0)
```
Comment (dlech) [inline lib/pbio/test/pbsysconfig.h:7]: It would be nice to test this code.

**Review by dlech — COMMENTED** [review]
Nice work. I made a bunch of comments about all of the little details, but overall it looks like a nice improvement. The user program memory allocation code is a bit hard to understand so I would need more time to really go through it to understand it enough to be able to spot any potential problems or offer suggestions for improvements there. Likewise, the startup and shutdown sequencing really needs some more thought to ensure correctness.

File: bricks/cityhub/city_hub.ld [hunk 3]
```diff
@@ -9,11 +9,16 @@
...
-    /* Flash size is 256K, bootloader resides in first 20K, last 4K seems to be reserved */
+    /* Flash size is 256K, bootloader resides in first 20K. */
     FLASH (rx)      : ORIGIN = 0x08005000, LENGTH = 232K
     USER_FLASH (rx) : ORIGIN = 0x0803F000, LENGTH = 4K
     RAM (xrw)       : ORIGIN = 0x20000000, LENGTH = 32K
 }
 
 "MAGIC_OFFSET" = 0x100;
-_minimum_stack_size = 2K;
+_stack_size = 4K;
```
[the remark this answers is not in the document]
Comment (laurensvalk) [reply bricks/cityhub/city_hub.ld:19]: Will do. This value was set to provide the same amount available today (the left over after setting a fixed heap size).

File: bricks/cityhub/Makefile [hunk 4]
```diff
@@ -6,6 +6,7 @@ PB_MCU_SERIES = F0
 PB_CMSIS_MCU = STM32F030xC
 PB_MCU_EXT_OSC_HZ = 0 # uses internal oscillator
 PB_FIRMWARE_MAX_SIZE = 237568
+PB_USE_HAL = 1
```
[the remark this answers is not in the document]
Comment (laurensvalk) [reply bricks/cityhub/Makefile:9]: It's used to save the program on shutdown.

File: bricks/stm32/common.ld [hunk 5]
```diff
@@ -65,14 +66,25 @@ SECTIONS
         *(.bss*)
         *(COMMON)
 
-        . = ALIGN(4);
+        . = ALIGN(2048);
```
[the remark this answers is not in the document]
Comment (laurensvalk) [reply bricks/stm32/common.ld:69]: Thanks. Looks like one of my debugging tests slipped through; it's not meant to be this big.

**Comment by laurensvalk** [comment]
Thanks for the detailed review! I'll try going through all of them by the end of the week.

I've just added the final feature for this branch, so now it is time to begin cleaning it all up.

**Comment by laurensvalk** [comment]
I've just added the final major feature of this branch. This adds support for **multi-file projects**!

A script can now import a module from the same directory. This allows users to manage and re-use frequently used code snippets and classes.

An example implementation for `pybricksdev` is given in https://github.com/pybricks/pybricksdev/pull/37. It essentially checks recursively for dependencies of the main script, and bundles them. This should be feasible in Pybricks Code as well.

On the hub itself, they are treated as real modules, so import statements work without special behaviors.

The exact format of the bundled code is up for discussion. The currently implemented format is:
```
    The returned bytes format is of the form:

       - first script size (uint32 little endian)
       - first script name (no extension, zero terminated string)
       - first script mpy data
       - second script size
       - second script name
       - second script mpy data
       - ...

    The first script is assumed to be the main script. The order of the other scripts (the modules) is arbitrary.

    All components are 0-padded to a size with a multiple of 4.

    If the main script does not import any local module, it returns only the
    first script mpy data (without size and name) for backwards compatibility
    with older firmware.
```

File: lib/pbio/sys/bluetooth.c [hunk 6]
```diff
@@ -234,7 +235,8 @@ PROCESS_THREAD(pbsys_bluetooth_process, ev, data) {
     pbdrv_bluetooth_set_on_event(on_event);
     pbdrv_bluetooth_set_receive_handler(handle_receive);
 
-    while (!pbsys_status_test(PBIO_PYBRICKS_STATUS_SHUTDOWN)) {
+    while (1
```
[the remark this answers is not in the document]
Comment (laurensvalk) [reply lib/pbio/sys/bluetooth.c:238]: This is just a hack to keep bluetooth enabled during shutdown for debugging purposes. There is a to-do note in this PR about removing it when everything is done.

File: lib/pbio/sys/main.c [hunk 7]
```diff
@@ -1,47 +1,86 @@
...
 }
 
+void pbsys_init() {
+    pbsys_battery_init();
+    pbsys_bluetooth_init();
+    pbsys_hmi_init();
+    process_start(&pbsys_system_process);
+}
+
+#if PBSYS_CONFIG_PROGRAM_LOAD
```
[the remark this answers is not in the document]
Comment (laurensvalk) [reply lib/pbio/sys/main.c:58]: The virtual hub doesn't use it.

File: lib/pbio/sys/main.c [hunk 8]
```diff
@@ -1,47 +1,86 @@
...
-        // REVISIT: we could save a few CPU cycles on each call to pbio_do_one_event()
-        // if we don't set this until shutdown is actually requested
-        pbio_set_event_hook(pb_sys_main_check_for_shutdown);
-        main();
-    } else {
-        // in case we jumped out of the middle of a user program
-        pbsys_user_program_unprepare();
+        // Wait for program receive process to complete.
+        while (!pbsys_program_load_receive_complete()) {
+            pbio_do_one_event();
```
[the remark this answers is not in the document]
Comment (laurensvalk) [reply lib/pbio/sys/main.c:74]: There already is; the process stops on shutdown.

File: bricks/essentialhub/essential_hub.ld [hunk 9]
```diff
@@ -16,4 +16,11 @@ MEMORY
...
 "FW_INFO_OFFSET" = 0x200;
-_minimum_stack_size = 8K;
+_stack_size = 16K;
+
+/* On this hub, user data is stored on the external flash, so it does not
+ * directly affect the internal flash layout. We still define these symbols
+ * here so we can ensure there is enough available RAM to load the program.
+ */
+_app_data_flash_start = 512K;
+_app_data_flash_end = _app_data_flash_start + 256K;
```
[the remark this answers is not in the document]
Comment (laurensvalk) [reply bricks/essentialhub/essential_hub.ld:26]: Linker files can't read the config headers though. This is now handled differently so marking as resolved.

**Comment by laurensvalk** [comment]
This has been split out to #109 (download and run), #110 (non-blocking spi), #112 (powered up storage), and https://github.com/pybricks/pybricks-micropython/tree/fix-526, so we can close this one.

**Pull request closed without merging**